lib/axon.ex

```diff
@@ -3495,8 +3487,36 @@ defmodule Axon do
   or backend.
   """
   @doc type: :compilation
-  def compile(model, opts \\ []) when is_list(opts) do
-    {Axon.Compiler.compile_init(model, opts), Axon.Compiler.compile_predict(model, opts)}
+  def build(model, opts \\ []) when is_list(opts) do
+    {init_fn, predict_fn} = Axon.Compiler.build(model, opts)
+    opts = [on_conflict: :reuse] ++ opts
+    # TODO: Wrap in jit_apply
+    {&Nx.Defn.jit(init_fn, [&1, &2], opts), &Nx.Defn.jit(predict_fn, [&1, &2], opts)}
+  end
+
+  @doc """
+  Compiles the given model to `{init_fn, predict_fn}`.
+
+  This function will compile a model specialized to the given
+  input shapes and types. This is useful for avoiding the overhead
+  of long compilations at program runtime. You must provide template
+  inputs which match the expected shapes and types of inputs at
+  execution time.
+
+  This function makes use of the built-in `Nx.Defn.compile/3`. Note
+  that passing inputs which differ in shape or type from the templates
+  provided to this function will result in potentially expensive
+  recompilation.
+  """
+  @doc type: :compilation
+  def compile(model, template, init_params \\ %{}, opts \\ []) when is_list(opts) do
```

Comment:
I am afraid this won't play well with optional model parameters. When you have optional model parameters, then you need to train the function for those precise model parameters and the default model state won't have them, correct?

I don't even know if `init_params` itself can have optional parameters... but we may consider the fact we only want to compile `predict`?

```elixir
def compile(model, template, opts \\ [])

def compile(model, template, opts) when is_list(opts) do
  compile(model, template, nil, opts)
end

def compile(model, template, model_params) when is_map(model_params) do
  compile(model, template, model_params, [])
end

def compile(model, template, model_params, opts) do
  {init_fn, predict_fn} = build(model, opts)
  Nx.Defn.compile(predict_fn, [model_params || init_fn.(template, init_params), template])
end
```

Reply:
The code above obviously won't work, because we don't have `init_params` on the fallback case. So maybe we can't allow compiling without the model parameters at all.

Reply:
I am 100% wrong. Just ignore me. :)